[AArch64] Combine PTEST_FIRST(PTRUE, CONCAT(A, B)) -> PTEST_FIRST(PTRUE, A)

kmclaughlin-arm opened the pull request:
When the input to ptest_first is a vector concat and the mask is all active,
performPTestFirstCombine returns a ptest_first using the first operand
of the concat, looking through any reinterpret casts.

This allows optimizePTestInstr to later remove the ptest when the first
operand is a flag setting instruction such as whilelo.

kmclaughlin-arm committed:
[AArch64] Combine PTEST_FIRST(PTRUE, CONCAT(A, B)) -> PTEST_FIRST(PTRUE, A)

When input to a ptest_first is a vector concat and the mask is all
active, performPTestFirstCombine returns a ptest_first using the
first operand of the concat, looking through any reinterpret casts
added by getPTest.

This allows optimizePTestInstr to later remove the ptest when the
first operand is a flag setting instruction such as whilelo.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -27519,6 +27522,37 @@ static SDValue performMULLCombine(SDNode *N,
+    Pred = Pred->getOperand(0);
+    SDValue Mask = DAG.getSplatVector(Pred->getValueType(0), DL,
+                                      DAG.getAllOnesConstant(DL, MVT::i64));
+    return getPTest(DAG, N->getValueType(0), Mask, Pred,
+                    AArch64CC::FIRST_ACTIVE, /* EmitCSel */ false);
```
I doubt you're using much in `getPTest` to warrant the `EmitCSel` change. Having already proven the DAG is testing the first lane of the predicate[1] there's no complexity related to "hidden lanes" and so this can be simplified to:
```
Pred = DAG.getNode(AArch64ISD::REINTERPRET_CAST, DL, nxv16i1, Pred)
Mask = DAG.getNode(AArch64ISD::REINTERPRET_CAST, DL, nxv16i1, Mask)
return DAG.getNode(AArch64ISD::PTEST_FIRST, DL, N->getValueType(0), Pred, Mask);
```

[1] Is worth adding commentary detailing the "we know we are testing the first lane" requirement because that's key to the combine.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -27519,6 +27522,37 @@ static SDValue performMULLCombine(SDNode *N,
+  if (Pred->getOpcode() == AArch64ISD::REINTERPRET_CAST)
+    Pred = Pred->getOperand(0);
+
+  if (Pred->getValueType(0).getVectorElementType() != MVT::i1 ||
```
You don't need the element type check because `AArch64ISD::REINTERPRET_CAST` is only allowed to change the element count.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -27519,6 +27522,37 @@ static SDValue performMULLCombine(SDNode *N,
+
+  if (Pred->getOpcode() == ISD::CONCAT_VECTORS) {
+    Pred = Pred->getOperand(0);
+    SDValue Mask = DAG.getSplatVector(Pred->getValueType(0), DL,
+                                      DAG.getAllOnesConstant(DL, MVT::i64));
```
You've already proven the existing `Mask` does what you need so should be reusable.  Perhaps it's worth adding an `isLanes1toNKnownOne` helper function, then you don't need to strip reinterpret from `Mask` and can then reuse it directly.

kmclaughlin-arm committed:
- Add isLane1KnownActive helper
- Extend canRemovePTestInstr for PTEST_PP_FIRST
- Remove getVectorElementType() != MVT::i1 check for Mask

kmclaughlin-arm committed:
- Rename isLane1KnownActive -> isLane0KnownActive

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -27519,6 +27534,32 @@ static SDValue performMULLCombine(SDNode *N,
+  if (Pred->getOpcode() == AArch64ISD::REINTERPRET_CAST)
+    Pred = Pred->getOperand(0);
+
+  if (!isLane0KnownActive(Mask))
+    return SDValue();
```
Perhaps move this before looking through Pred's reinterpret_cast because this is what makes that code logically safe.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -1503,6 +1503,10 @@ AArch64InstrInfo::canRemovePTestInstr(MachineInstr *PTest, MachineInstr *Mask,
             getElementSizeForOpcode(PredOpcode))
       return PredOpcode;
 
+    if (PTest->getOpcode() == AArch64::PTEST_PP_FIRST &&
```
Perhaps add a comment to match the other cases?

kmclaughlin-arm committed:
- Move isLane0KnownActive before looking through reinterpret_cast
 - Add comment to changes in canRemovePTestInstr
